### PR TITLE
[Merged by Bors] - `bevy_ptr` standalone crate

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

The pointer types introduced in #3001 are useful not just in `bevy_ecs`, but also in crates like `bevy_reflect` (#4475) or even outside of bevy.

## Solution

Extract `Ptr<'a>`, `PtrMut<'a>`, `OwnedPtr<'a>`, `ThinSlicePtr<'a, T>` and `UnsafeCellDeref` from `bevy_ecs::ptr` into `bevy_ptr`.

**Note:** `bevy_ecs` still reexports the `bevy_ptr` as `bevy_ecs::ptr` so that crates like `bevy_transform` can use the `Bundle` derive without needing to depend on `bevy_ptr` themselves.

### CODE REVIEW

File: crates/bevy_ptr/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "bevy_ptr"
+version = "0.8.0-dev"
+edition = "2021"
+description = "Bevy Engine's entity component system"
```
Comment (mockersf) [inline crates/bevy_ptr/Cargo.toml:5]: Could you change the description?

**Comment by alice-i-cecile** [comment]
Very reasonable. I definitely think these types are widely useful, and I'd enjoy being able to contribute back to the broader ecosystem where we can.

@cart if you're on board with this remember to make an `A-Pointers` label :)

**Comment by mockersf** [comment]
Could you add the new crate to https://github.com/bevyengine/bevy/blob/main/tools/publish.sh?

**Review by bjorn3 — APPROVED** [review]

**Review by TheRawMeatball — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Review by IceSentry — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
> Also, is this change really that controversial?

We still need to have a high-level discussion about our crate splitting strategy. That said, I'm going to add this to https://github.com/orgs/bevyengine/projects/6/views/1 now that Francois has approved it.

**Pull request closed without merging**